File: 2021/day2/main.py
import sys

# aim -> 0
# down X -> increases aim by X units
# up X -> decreases aim by X units
# forward X ->
#   increases horizontal position by X units
#   increases depth by (aim * X units)


def main() -> None:
    AIM = 0
    DEPTH = 0
    HORIZONTAL = 0

    with open("src/input.txt", "r") as f:
        d = f.read()

        l = d.splitlines()

        for line in l:
            cmd, val = line.split()

            match cmd:
                case "forward":
                    HORIZONTAL += int(val)
                    DEPTH += AIM * int(val)
                case "down":
                    AIM += int(val)
                case "up":
                    AIM -= int(val)
                case _:
                    raise ValueError(f"unknown_command={cmd}")

    print(f"result={DEPTH * HORIZONTAL}")


if __name__ == "__main__":
    sys.exit(main())

File: 2021/day2/src/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::str::FromStr;

// forward u8
// down u8
// up u8
enum Command {
    Forward(u8),
    Down(u8),
    Up(u8),
}

#[derive(Debug)]
struct ParseCommandError;

impl std::fmt::Display for ParseCommandError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "failed to parse command")
    }
}

impl std::error::Error for ParseCommandError {}

impl FromStr for Command {
    type Err = ParseCommandError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (cmd, arg) = s.split_once(' ').ok_or(ParseCommandError)?;
        let value = arg.parse().map_err(|_| ParseCommandError)?;

        match cmd {
            "forward" => Ok(Command::Forward(value)),
            "down" => Ok(Command::Down(value)),
            "up" => Ok(Command::Up(value)),
            _ => Err(ParseCommandError),
        }
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open("input.txt")?;
    let reader = BufReader::new(file);

    let mut horizontal: u16 = 0;
    let mut depth: u16 = 0;

    let final_result: u32;

    for line in reader.lines() {
        let line = line?;
        let command: Command = line.parse()?;

        match command {
            Command::Forward(x) => horizontal += x as u16,
            Command::Down(x) => depth += x as u16,
            Command::Up(x) => depth -= x as u16,
        }
    }

    final_result = horizontal as u32 * depth as u32;
    println!("final_result={}", final_result);

    Ok(())
}
